docs/user_guide/proxy.md

```diff
@@ -0,0 +1,94 @@
+# Use dfdaemon as an HTTP proxy
+
+Dfdaemon can be used as an HTTP proxy to speed up image pulling from any registry
+as well as general HTTP downloads.
+
+Please first ensure that you know how to install and run [supernode](install_server.md)
+and [dfdaemon](install_client.md).
+
+## Download images
+
+Add the following content to `/etc/dragonfly/dfdaemon.yml`.
+
+```yaml
+proxies:
+# proxy all http image layer download requests with dfget
+- regx: blob/sha256/.*
+```
+
+Set HTTP_PROXY for docker daemon in `/etc/systemd/system/docker.service.d/http-proxy.conf`.
+`65002` is the default proxy port for dfdaemon. This can be configured with the
+`--proxyPort` cli parameter of dfdaemon.
+
+```
+[Service]
+Environment="HTTP_PROXY=http://127.0.0.1:65002"
+```
+
+Currently dfdaemon can't hijack HTTPS requests, so if you want dragonfly to speed
+up image download in a registry, you have to set it as insecure in `/etc/docker/daemon.json`.
+But rest assured that the content of authentication requests will still be encrypted.
+
+```json
+{
+  "insecure-registries": [ "your.registry" ]
+}
+```
+
+This requires that your registry can handle HTTP requests for image layers,
```

Comment:
@inoc603 FYI.

```
With insecure registries enabled, Docker goes through the following steps:

First, try using HTTPS.
If HTTPS is available but the certificate is invalid, ignore the error about the certificate.
If HTTPS is not available, fall back to HTTP.
```

And I have tried it locally and it couldn't support HTTPS registry in this way. So could you please show me something that prove you downloaded an image from HTTPS registry successfully? THX. 😄

Reply:
Thanks for point that out, I find that although the registry I used to test enabled https, but it uses oss for image storage and the image layer urls it returns are http. I thought it was the `insecure-registries` option that made all the url http, clearly it was a mistake.

If the registry uses different domain names for authentication and image download, it would be easy. However that may not always be the case, if it has only one domain, currently I don't have a simple way to tell a auth request from a image download request if they are https.

Reply:
But I do suggest that we clearly separate the usage of `port` and `proxyPort`. Use `port` for reverse proxy or `registry-mirrors`, and use `proxyPort` for `HTTP_PROXY`

Reply:
We can probably use a local CA to decrypt https traffic, but that's kind of hacky. The current https support is limited to "not breaking it", but does not speed them up with dfget.

Reply:
And I think if we manage to implement the man-in-the-middle https proxy, it should be optional, and we'll have different level of https support for the users:
- only http are supported, https gets blocked
- https gets tunneled directly
- https gets decrypted and speed up with dfget

Reply:
> We can probably use a local CA to decrypt https traffic, but that's kind of hacky. The current https support is limited to "not breaking it", but does not speed them up with dfget.

We have considered this option, but it is not an ideal approach.

Reply:
> We can probably use a local CA to decrypt https traffic, but that's kind of hacky. The current https support is limited to "not breaking it", but does not speed them up with dfget.

We have considered this approach before, and if there is really no better way, I think we have to use this method. 😭

Reply:
> But I do suggest that we clearly separate the usage of `port` and `proxyPort`. Use `port` for reverse proxy or `registry-mirrors`, and use `proxyPort` for `HTTP_PROXY`

I agree with you. However, it will make uses feel confused. What do you think of that  we use `port` to for  images and use `proxyPort` for ordinary files?

Reply:
> And I think if we manage to implement the man-in-the-middle https proxy, it should be optional, and we'll have different level of https support for the users:
> 
> * only http are supported, https gets blocked
> * https gets tunneled directly
> * https gets decrypted and speed up with dfget

I will submit an issue later,  we can do more discussion at that time. And I'll let you know at the first time.

Reply:
> > But I do suggest that we clearly separate the usage of `port` and `proxyPort`. Use `port` for reverse proxy or `registry-mirrors`, and use `proxyPort` for `HTTP_PROXY`
> 
> I agree with you. However, it will make users feel confused. What do you think of that we use `port` to for images and use `proxyPort` for ordinary files?

Actually we can use the same port for both, but that would require some changes on the current http api, because we can't handle CONNECT requests with the current `mux` based approach. I didn't want to change the api to much at the time I implement the transparent proxy, so I used a new port, but it's certainly possible to do so.